### PR TITLE
feat(totp): totp client-side management apis

### PR DESCRIPTION
**Pull request by vbudhram**
Connects to https://github.com/mozilla/fxa-content-server/issues/5863

A WIP PR for client side functions needed for TOTP.

### CODE REVIEW

**Comment by vbudhram** [comment]
@mozilla/fxa-devs This is ready for review, if you get a chance.

**Comment by rfk** [comment]
Sorry for the merge conflict, but as penance, I'm going to review this right now :-)

File: package.json [hunk 1]
```diff
@@ -53,6 +53,7 @@
     "intern-geezer": "2.2.3",
     "jscs-jsdoc": "1.1.0",
     "load-grunt-tasks": "3.2.0",
+    "otplib": "7.1.0",
```
Comment (rfk) [inline package.json]: To sanity-check my understanding, we need this in order to run the tests against a live server, is that accurate?

File: tests/lib/totp.js [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+], function (tdd, assert, Environment, sinon, otplib) {
+
+  with (tdd) {
+    suite('totp', function () {
+      var account;
+      var accountHelper;
+      var respond;
+      var client;
+      var RequestMocks;
+      var env = new Environment();
```
Comment (rfk) [inline tests/lib/totp.js]: Is it necessary to initialize `env` here, if we assign it a value in `beforeEach`?

File: tests/lib/totp.js [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+      beforeEach(function () {
+        env = new Environment();
+        accountHelper = env.accountHelper;
+        respond = env.respond;
+        client = env.client;
+        RequestMocks = env.RequestMocks;
+
+        return accountHelper.newVerifiedAccount()
+          .then(function (newAccount) {
+            account = newAccount;
```
Comment (rfk) [inline tests/lib/totp.js]: nit: this local variable seems a little redundant, we could just name the parameter `account`.

File: tests/lib/totp.js [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+            return respond(client.createTotpToken(account.signIn.sessionToken), RequestMocks.createTotpToken);
+          })
+          .then(function (res) {
+            assert.ok(res.qrCodeUrl, 'should return QR code data encoded url');
+            assert.ok(res.secret, 'should return secret that is encoded in url');
+
+            otplib.authenticator.options = {
+              secret: res.secret
+            };
```
Comment (rfk) [inline tests/lib/totp.js]: IIUC this `otplib.authenticator` is mutable state shared between tests; could we use ` new otplib.authenticator.Authenticator()` to ensure that the tests don't accidentally share state?

File: package.json (hunk 1 above)
Comment (vbudhram) [reply package.json]: That is correct, it is able to generate a valid code from the secret that is return from the server.

File: tests/lib/totp.js (hunk 3 above)
Comment (vbudhram) [reply tests/lib/totp.js]: This is actually setting the value for the global [account object](https://github.com/mozilla/fxa-js-client/blob/ec00e34a8438a2e15098d0bcfc961b65fb3bb6fe/tests/lib/totp.js#L15), which is reused in each test case.

Comment (rfk) [reply tests/lib/totp.js]: Oh right, of course :-)

**Comment by vbudhram** [comment]
@rfk Thanks! I believe this should be last item for train 107 in this repo.

**Pull request merged**